todo_completion

```diff
@@ -31,51 +31,53 @@ _todo()
-                            )
-                            # Append task text as a shell comment. This
-                            # completion can be a safety check before a
-                            # destructive todo.txt operation.
-                            [ "$todo" ] && COMPREPLY[0]="$cur # $todo"
-                            return 0
-                        else
-                            return 0
-                        fi
-                        ;;
-                esac
+            *)  if [[ "$prev" =~ ^(addto|listfile|lf${_todo_file1_actions:+|${_todo_file1_actions}})$ ]]; then
+                    completions=$(eval TODOTXT_VERBOSE=0 $_todo_sh command listfile 2>/dev/null)
+                else
+                    case "$cur" in
+                        +*) completions=$(eval TODOTXT_VERBOSE=0 $_todo_sh command listproj 2>/dev/null)
+                            COMPREPLY=( $( compgen -W "$completions" -- $cur ))
+                            [ ${#COMPREPLY[@]} -gt 0 ] && return 0
+                            # Fall back to projects extracted from done tasks.
+                            completions=$(eval 'TODOTXT_VERBOSE=0 TODOTXT_SOURCEVAR=\$DONE_FILE' $_todo_sh command listproj 2>/dev/null)
+                            ;;
+                        @*) completions=$(eval TODOTXT_VERBOSE=0 $_todo_sh command listcon 2>/dev/null)
+                            COMPREPLY=( $( compgen -W "$completions" -- $cur ))
+                            [ ${#COMPREPLY[@]} -gt 0 ] && return 0
+                            # Fall back to contexts extracted from done tasks.
+                            completions=$(eval 'TODOTXT_VERBOSE=0 TODOTXT_SOURCEVAR=\$DONE_FILE' $_todo_sh command listcon 2>/dev/null)
+                            ;;
+                        *)  if [[ "$cur" =~ ^[0-9]+$ ]]; then
+                                # Remove the (padded) task number; we prepend the
+                                # user-provided $cur instead.
+                                # Remove the timestamp prepended by the -t option,
+                                # and the done date (for done tasks); there's no
+                                # todo.txt option for that yet.
+                                # But keep priority and "x"; they're short and may
+                                # provide useful context.
+                                # Remove any trailing whitespace; the Bash
+                                # completion inserts a trailing space itself.
+                                # Finally, limit the output to a single line just as
+                                # a safety check of the ls action output.
+                                local todo=$( \
+                                    eval TODOTXT_VERBOSE=0 $_todo_sh '-@ -+ -p -x command ls "^ *${cur} "' 2>/dev/null | \
+                                    sed -e 's/^ *[0-9]\{1,\} //' -e 's/^\((.) \)\{0,1\}[0-9]\{2,4\}-[0-9]\{2\}-[0-9]\{2\} /\1/' \
```

Comment:
I know it would be _slower_, but would it be wise to split up the `sed` command into separate calls with the comments above nearby. I believe it would be easier to maintain.

Reply:
@karbassi Yes, that long comment block is ugly. It's all existing code, though; this patch just changed the indentation level, that's why it shows up in the diff. It would be unfair to blame this patch for it.

I've just tried to improve this in #273; after all, it was me who initially created that mess :-)